src/prefect/events/related.py

```diff
@@ -0,0 +1,94 @@
+from typing import TYPE_CHECKING, Dict, List, Optional, Set, Tuple
+from uuid import UUID
+
+from .schemas import RelatedResource
+
+if TYPE_CHECKING:
+    from prefect.client.schemas import FlowRun
+    from prefect.server.schemas.core import Flow
+
+related_resource_cache: Dict[UUID, Tuple[Optional["FlowRun"], Optional["Flow"]]] = {}
+
+
+async def related_resources_from_run_context(
+    exclude: Optional[Set[str]] = None,
+) -> List[RelatedResource]:
+    from prefect.context import FlowRunContext, TaskRunContext
+
+    if exclude is None:
+        exclude = set()
+
+    flow_run_id: Optional[UUID] = None
+    flow_run: Optional["FlowRun"] = None
+    flow: Optional["Flow"] = None
+
+    flow_run_context = FlowRunContext.get()
+
+    if flow_run_context is None:
+        task_run_context = TaskRunContext.get()
+        if task_run_context is None:
+            return []
+
+        flow_run_id = task_run_context.task_run.flow_run_id
+
+        if flow_run_id in related_resource_cache:
+            flow_run, flow = related_resource_cache[flow_run_id]
+        else:
+            client = task_run_context.client
```

Comment:
It's risky to grab the client off the context like this fyi it'll only work if you're on the same event loop as the client was created in. It's okay in this case but may not be in the future.